NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+
+  symmetrytable_ = BuildSymmetryTable();
+
+  colors_.reserve(nlatticesites_);
+  for (auto const &edge : edges_) {
+    auto success = colors_.emplace(edge, 0).second;
+    static_cast<void>(success);  // Make everyone happy in the NDEBUG case
+    assert(success && "There should be no duplicate edges");
+  }
+
+  is_connected_ = ComputeConnected();
+  is_bipartite_ = ComputeBipartite();
+
+};  // namespace netket
+Lattice::~Lattice(){};
+
+// Get private member
+int Lattice::Ndim() const noexcept { return ndim_; };
+int Lattice::Nsites() const noexcept { return nsites_; };
+int Lattice::Size() const noexcept { return Nsites(); };
+
+std::vector<std::vector<double>> Lattice::Coordinates() const noexcept {
+  std::vector<std::vector<double>> result;
+  for (int k = 0; k < nsites_; k++) result.push_back(R_[k]);
+  return result;
+}
+std::vector<Lattice::Edge> const &Lattice::Edges() const noexcept {
+  return edges_;
+};
+std::vector<std::vector<int>> Lattice::SymmetryTable() const noexcept {
+  return symmetrytable_;
+};
+
+// Returns map of the edge and its respective color
+const AbstractGraph::ColorMap &Lattice::EdgeColors() const noexcept {
+  return colors_;
+}
+
+bool Lattice::ComputeConnected() const {
+  const int start = 0;  // arbitrary node
+  int nvisited = 0;
+  BreadthFirstSearch(start, [&nvisited](int, int) { ++nvisited; });
+  return nvisited == Nsites();
+}
+
+bool Lattice::ComputeBipartite() const {
+  bool is_bipartite = true;
+  const int start = 0;  // arbitrary node
+  std::vector<int> colors(Nsites(), -1);
+  const auto adjacency_list =
+      AdjacencyList();  // implicit expression can't have
+                        // access to the Lattice function
+  if (is_connected_) {
+    BreadthFirstSearch(
+        start, [&colors, &adjacency_list, &is_bipartite](int node, int) {
+          if (node == start) colors[node] = 1;
+          for (std::size_t j = 0; j < adjacency_list[node].size(); j++) {
+            if (!is_bipartite) break;
+            if (colors[adjacency_list[node][j]] == -1) {
+              colors[adjacency_list[node][j]] = 1 - colors[node];
+            } else if (colors[adjacency_list[node][j]] == colors[node]) {
+              is_bipartite = false;
+            }
+          }
+        });
+  } else {
+    BreadthFirstSearch(
+        [&colors, &adjacency_list, &is_bipartite](int node, int, int) {
+          if (node == start) colors[node] = 1;
+          for (std::size_t j = 0; j < adjacency_list[node].size(); j++) {
+            if (!is_bipartite) break;
+            if (colors[adjacency_list[node][j]] == -1) {
+              colors[adjacency_list[node][j]] = 1 - colors[node];
+            } else if (colors[adjacency_list[node][j]] == colors[node]) {
+              is_bipartite = false;
+            }
+          }
+        });
+  }
+  return is_bipartite;
+}
+
+// Lattice representations (site = k, vector = n_i, coord = cartesian
+// coordinates)
+std::vector<int> Lattice::Site2Vector(int i) const {

```

Comment:
Could you please add an assert that `i >= 0`?

Reply:
Done